.ci/azure-pipelines/azure-pipelines.yaml

```diff
@@ -1,163 +1,85 @@
 
 pr:
   paths:
     exclude:
-    - doc
     - README.md
     - CHANGES.md
     - CONTRIBUTING.md
+    # add more exclude rules, e.g. .ci, .dev, .github
+    # which won't cause any CI to run.
 
 resources:
   containers:
-    - container: fmt
+    - container: fmt # for formatting.yaml
       image: pointcloudlibrary/fmt
-    - container: env1804
-      image: pointcloudlibrary/env:18.04
-    - container: env2004
-      image: pointcloudlibrary/env:20.04
-    - container: env2010
-      image: pointcloudlibrary/env:20.10
 
 stages:
   - stage: formatting
     displayName: Formatting
     jobs:
       - template: formatting.yaml
 
-  - stage: build_gcc
-    displayName: Build GCC
-    dependsOn: formatting
+  - stage: diff
+    displayName: Code Changes Diff
     jobs:
-      - job: ubuntu
-        displayName: Ubuntu
+      - job: codeDiff
+        displayName: Code Diff
         pool:
-          vmImage: 'Ubuntu 20.04'
-        strategy:
-          matrix:
-            18.04 GCC:  # oldest LTS
-              CONTAINER: env1804
-              CC: gcc
-              CXX: g++
-              BUILD_GPU: ON
-              CMAKE_ARGS: '-DPCL_WARNINGS_ARE_ERRORS=ON'
-            20.10 GCC:  # latest Ubuntu
-              CONTAINER: env2010
-              CC: gcc
-              CXX: g++
-              BUILD_GPU: OFF
-              # surface is not ready for re-entrant QHull
-              CMAKE_ARGS: '-DBUILD_tests_surface=OFF'
-        container: $[ variables['CONTAINER'] ]
-        timeoutInMinutes: 0
-        variables:
-          BUILD_DIR: '$(Agent.BuildDirectory)/build'
-          CMAKE_CXX_FLAGS: '-Wall -Wextra -Wnoexcept-type'
-          DISPLAY: :99.0 # Checked for in CMake
+          vmImage: 'ubuntu-latest'
         steps:
-          - template: build/ubuntu.yaml
+          - checkout: self
+          - script: | 
+              # Add all docs path to this array
+              DOCS_LISTS=("doc" "*.md")
+              for ENTRY in "${DOCS_LISTS[@]}"; do
+                DIFF_EXCLUDE="':(exclude)${ENTRY}' ${DIFF_EXCLUDE}"
+              done
+              cd $(Build.SourcesDirectory)
+              echo "##vso[task.setvariable variable=HasCodeChanges;isOutput=true]false"
+              # Show full git diff results
+              git diff $(git merge-base HEAD origin/test-CI-features)
+              # Set HasCodeChanges to true, if there's still file changed when excluding those docs
+              git diff --quiet $(git merge-base HEAD origin/test-CI-features) -- ${DIFF_EXCLUDE} || \
+              echo "##vso[task.setvariable variable=HasCodeChanges;isOutput=true]true"
+            name: GitResult
+            displayName: Show Git Diff Result
 
-  - stage: build_clang
-    displayName: Build Clang
-    dependsOn: formatting
+  - stage: documentation
+    displayName: Goto Documentation
+    dependsOn: diff
+    condition: and(succeeded(), eq(dependencies.diff.outputs['CodeDiff.GitResult.HasCodeChanges'], 'false'))
```

Comment:
This seems weird. If a md file is changed, documentation will be run. Perhaps, add separate variables for doc and build stages

Reply:
Thanks for your advice! Although `*.md` is a part of `DOCS_LISTS`, it doesn't mean md file would trigger CI(the path-filter will restrict md files from running any CI), and it's just used to separate the code and docs changes(i.e. I assume changes will only be code or docs changes).

Defining changes as **codes**, **docs** or **others** changes is undoubtedly an excellent idea to improve readability. But it would make the script a lit complex, should we do that in the next step?